mono/metadata/loaded-images.c

```diff
@@ -0,0 +1,101 @@
+#include "config.h"
+
+#include "mono/metadata/loaded-images-internals.h"
+#include "mono/metadata/metadata-internals.h"
+#include "mono/utils/mono-logger-internals.h"
+
+void
+mono_loaded_images_init (MonoLoadedImages *li, MonoAssemblyLoadContext *owner)
+{
+	li->owner = owner;
+	for (int hash_idx = 0; hash_idx < MONO_LOADED_IMAGES_HASH_COUNT; hash_idx++)
+		li->loaded_images_hashes [hash_idx] = g_hash_table_new (g_str_hash, g_str_equal);
+}
+
+void
+mono_loaded_images_cleanup (MonoLoadedImages *li, gboolean shutdown)
+{
+	if (shutdown) {
+		GHashTableIter iter;
+		MonoImage *image;
+
+		// If an assembly image is still loaded at shutdown, this could indicate managed code is still running.
+		// Reflection-only images being still loaded doesn't indicate anything as harmful, so we don't check for it.
+		g_hash_table_iter_init (&iter, mono_loaded_images_get_hash (li, FALSE));
+		while (g_hash_table_iter_next (&iter, NULL, (void**)&image))
+			mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_ASSEMBLY, "Assembly image '%s' [%p] still loaded at shutdown.", image->name, image);
+	}
+
+	for (int hash_idx = 0; hash_idx < MONO_LOADED_IMAGES_HASH_COUNT; hash_idx++) {
+		g_hash_table_destroy (li->loaded_images_hashes [hash_idx]);
+		li->loaded_images_hashes [hash_idx] = NULL;
+	}
+}
+
+void
+mono_loaded_images_free (MonoLoadedImages *li)
+{
+	mono_loaded_images_cleanup (li, FALSE);
+	g_free (li);
+}
+
+GHashTable *
+mono_loaded_images_get_hash (MonoLoadedImages *li, gboolean refonly)
+{
+	g_assert (li != NULL);
+	GHashTable **loaded_images_hashes = &li->loaded_images_hashes[0];
+	int idx = refonly ? MONO_LOADED_IMAGES_HASH_PATH_REFONLY : MONO_LOADED_IMAGES_HASH_PATH;
+	return loaded_images_hashes [idx];
+}
+
+GHashTable *
+mono_loaded_images_get_by_name_hash (MonoLoadedImages *li, gboolean refonly)
+{
+	g_assert (li != NULL);
+	GHashTable **loaded_images_hashes = &li->loaded_images_hashes[0];
+	int idx = refonly ? MONO_LOADED_IMAGES_HASH_NAME_REFONLY : MONO_LOADED_IMAGES_HASH_NAME;
+	return loaded_images_hashes [idx];
+}
+
+static MonoLoadedImages *
+loaded_images_get_owner (MonoImage *image)
+{
+	/* image->alc could be NULL if we're closing an image that wasn't
+	 * registered yet (for example if two threads raced to open it and one
+	 * of them lost) */
+	MonoAssemblyLoadContext *alc = mono_image_get_alc (image);
+	return mono_alc_get_loaded_images (alc);
+}
+
+/* LOCKING: assumes the images lock is locked */
```

Comment:
I'm not sure I follow why the caller has to take the lock instead of doing it in this function. The only other thing it seems to be doing in the call site is modifying the refcount of the image, and that's even atomic.

Reply:
The issue is we need multiple things to happen together without another thread racing with us: 1. decrement the refcount down to zero, 2. remove the MonoImage from the hash tables.

The issue is that if another thread opens the same image after we decrement the refcount to == 0 and we continue to try to close it, the other thread will find the image in the hash and increment the refcount to > 0 but we will continue with unloading it because we think the refcount is == 0.

Another way to solve it would be to take the images lock here in this function and then check the refcount again, but that's not how it used to work, and I mostly just moved the code from `image.c` to this file (and changed it to use `MonoLoadedImages`)

Reply:
Gotcha. The alternative sounds preferable to me, mostly because I think it's worth minimizing functions that assume locks are already held (particularly non-static ones), but I'm fine either way. If you do keep it as-is, it would be nice if the comment was expanded a bit to justify the caller having to take the lock.

Reply:
I changed `mono_loaded_image_remove_image` to take the lock and to do the decrement and remove together.  It actually cleans up `mono_image_close_except_pools` nicely.  👍 Thanks for the suggestion!